File: ProblemSolve/24.cpp
/***********************
 * Swap Nodes in Pairs *
 ***********************/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    //현재 노드와 다음 노드를 스왑한다
    ListNode* swapPairs(ListNode* head) {
        if(head == NULL || head -> next == NULL){
            return head;
        }
        
        ListNode* temp;
        //다음 노드에 연결된 리스트 저장
        temp = head -> next;
        
        //재귀적으로 들어가 제일 뒤에서부터 스왑한다
        //제일 뒤에서부터 바꿔야 최종적으로 모두 바뀐 리스트가 된다
        //짝수번째 노드와 홀수번째 노드를 바꿔야한다. ex) 1 -> 2 -> 3 -> 4 => (2 -> 1) -> (4 -> 3)
        head -> next = swapPairs(head -> next -> next);
        
        //현재 노드를 다음 노드로, 다음 노드를 현재 노드로 바꾼다
        temp -> next = head;
        
        return temp;
    }
};

File: ProblemSolve/24.cs
/***********************
 * Swap Nodes in Pairs *
 ***********************/

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

public class Solution {
    public ListNode SwapPairs(ListNode head) {
        //끝에 하나거나 딱 맞아 떨어지면 그냥 리턴
        if(head == null || head.next == null){
            return head;
        }
        // 1 -> 2 -> 3 -> 4일때
        // ans = 2 -> 3 -> 4, head.next = 3 -> 4를 가지고 이 함수를 돈 결과
        // 3 -> 4 일때,
        // ans = 4, ans.next = head. 4 -> 3 => 스왑
        // 다시 위로 돌아가서 head = 1 -> 4 -> 3
        // ans = 2 -> 3 -> 4. ans.next = head => 2 -> 1 -> 4 -> 3
        ListNode ans = new ListNode();
        ans = head.next;
        head.next = SwapPairs(head.next.next);
        
        ans.next = head;                
        
        return ans;
    }
}